src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1190 @@
+            return;
+        
+        // We now know that the message handler is not attached by network, and that it is not a
+        // sendSynchronous() thread. So it must be a FredPluginFCPMessageHandler, and we determine
+        // what it is and dispatch the message to it.
+        final FredPluginFCPMessageHandler messageHandler
+            = (direction == SendDirection.ToServer) ? server.get() : client;
+
+        if(messageHandler == null) {
+            // server is a WeakReference which can be nulled if the server plugin was unloaded.
+            // client is not a WeakReference, we already checked for it to be non-null.
+            // Thus, in this case here, the server plugin has been unloaded so we can have
+            // an error message which specifically talks about the *server* plugin.
+            throw new IOException("The server plugin has been unloaded.");
+        }
+        
+        dispatchMessageLocallyToMessageHandler(messageHandler, direction, message);
+    }
+    
+    /**
+     * Backend for {@link #send(SendDirection, FCPPluginMessage)} to dispatch messages which need
+     * to be transported by network.<br><br>
+     * 
+     * This shall only be called for messages for which it was determined that the message handler
+     * is not a plugin running in the local VM.
+     */
+    private void dispatchMessageByNetwork(final SendDirection direction,
+            final FCPPluginMessage message)
+                throws IOException {
+        
+        // The message handler is attached by network.
+        // In theory, we could construct a mock FredPluginFCPMessagehandler object for it to
+        // pretend it was a local message. But then we wouldn't know the reply message immediately
+        // because the messages take time to travel over the network. This wouldn't work with the
+        // local message dispatching code as it needs to know the reply immediately so it can send
+        // it out. To get the reply, we would have to create a thread which would exist until the
+        // reply arrives over the network.
+        // So instead, for simplicity and reduced thread count, we just queue the message directly
+        // to the network queue here and return.
+        
+        assert (direction == SendDirection.ToClient)
+            : "By design, this class always shall execute in the same VM as the server plugin. "
+            + "So for networked messages, we should always be sending to the client.";
+        
+        assert (clientConnection != null)
+            : "Trying to send a message over the network to the client. "
+            + "So the network connection to it should not be null.";
+        
+        if (clientConnection.isClosed())
+            throw new IOException("Connection to client closed for " + this);
+        
+        clientConnection.outputHandler.queue(new FCPPluginServerMessage(serverPluginName, message));
```

Comment:
Fixed by commit c57abe82dae05e8a3a131ef5c6352d004426fda7 and 8d303650b29a1670e1cea0dd03466c2856058431.
Notice that there are probably hundreds of places in other fred code where this is also done wrong, which I did not fix. I added deprecation to encourage people to fix it in the future.